### PR TITLE
chore(suite-native): biometrics reducer

### PR DESCRIPTION
**Pull request by juriczech**
## Description

- Migrates Suite Native biometrics state from Jotai atoms to Redux so biometrics is managed consistently with the rest of the app.
- Adds a persisted `biometrics` Redux slice with selectors, thunks, and MMKV persistence for `isBiometricsEnabled`.
- Adds a migration from the legacy `isBiometricsOptionEnabled` Jotai storage value into Redux persistence.
- Reworks authentication, settings toggling, and app state handling through Redux thunks.
- Updates onboarding, settings, app freeze handling, and analytics to use Redux selectors.
- Removes obsolete Jotai biometrics hooks/storage and updates package dependencies.
- Big win is that now we can cover the logic with unit tests 🎉 

## 🌐 Preview deployments

🌐 **Suite Web preview:** [https://dev.suite.sldev.cz/suite-web/chore/native/biometrics-redux/web/](https://dev.suite.sldev.cz/suite-web/chore/native/biometrics-redux/web/)

## 🔍 Currents Test Results

🔍 **Suite desktop test results:** [View in Currents](https://app.currents.dev/projects/4ytF0E/runs?branch=chore/native/biometrics-redux&lookback=7)

🔍 **Suite web test results:** [View in Currents](https://app.currents.dev/projects/Og0NOQ/runs?branch=chore/native/biometrics-redux&lookback=7)

## 🔒 Quarantined E2E Tests

<details><summary>Trezor Suite (web) — 6 test(s)</summary>

| Test | Type |
|------|------|
| Account types suite > Add-account-types-non-BTC-coins | 🤖 auto |
| Account metadata > dropbox provider | 🤖 auto |
| Quarantine test: "Onboarding - create wallet,Success (basic)" | 🙋 manual |
| Quarantine test: "Database migration,Db migration between: release/22.5/web => develop/web" | 🙋 manual |
| Quarantine test: "Multiple sessions,Session overtaken by another" | 🙋 manual |
| Quarantine test: "Recovery T2T1 - dry run,Recovery after partial recovery" | 🙋 manual |

</details>

_Updated: 2026-06-03T12:33:57.168Z • 6 test(s) total_

<details><summary>Trezor Suite (desktop) — 4 test(s)</summary>

| Test | Type |
|------|------|
| Labeling migration > Migration from local file | 🤖 auto |
| Account types suite > Add-account-types-non-BTC-coins | 🤖 auto |
| Onboarding - create wallet > Success (basic) | 🤖 auto |
| Quarantine test: "Multiple sessions,Session overtaken by another" | 🙋 manual |

</details>

_Updated: 2026-06-03T12:30:41.736Z • 4 test(s) total_

### CODE REVIEW

File: suite-native/storage/src/tests/migrations/biometricsV1.test.ts
Comment (matejkriz) [inline suite-native/storage/src/tests/migrations/biometricsV1.test.ts:1]: wrong filename V2 vs V1

File: suite-native/storage/src/migrations/biometrics/v1.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    if (typeof storedValue !== 'boolean') {
+        return persistedState;
+    }
+
+    unecryptedJotaiStorage.remove(BIOMETRICS_PERSISTED_ATOM_STORAGE_KEY);
+
+    return {
+        ...persistedState,
+        isBiometricsEnabled: storedValue,
+    };
```
Comment (matejkriz) [inline suite-native/storage/src/migrations/biometrics/v1.ts:36]: In the initial migration it's better to return `undefined` instead of undefined oldState, than you don't need `as` at all:
```
export const migrateBiometricsAtomToRedux = (oldState: unknown) => {
    if (!oldState || !isPersistedState(oldState)) {
        return undefined;
    }

    const storedValue = parseStoredBoolean(
        unecryptedJotaiStorage.getString(BIOMETRICS_PERSISTED_ATOM_STORAGE_KEY),
    );

    if (typeof storedValue !== 'boolean') {
        return undefined;
    }

    unecryptedJotaiStorage.remove(BIOMETRICS_PERSISTED_ATOM_STORAGE_KEY);

    return {
        ...oldState,
        isBiometricsEnabled: storedValue,
    };
};
```

Comment (juriczech) [reply suite-native/storage/src/migrations/biometrics/v1.ts:36]: https://github.com/trezor/trezor-suite/pull/23224/commits/85b1d1bb602ace133d06f147a2ae912d797d5bc4

File: suite-native/storage/src/tests/migrations/biometricsV1.test.ts
Comment (juriczech) [reply suite-native/storage/src/tests/migrations/biometricsV1.test.ts:1]: https://github.com/trezor/trezor-suite/pull/23224/commits/85b1d1bb602ace133d06f147a2ae912d797d5bc4

File: suite-native/biometrics/src/components/BiometricsModalRenderer.tsx [hunk 2]
```diff
@@ -1,15 +1,48 @@
...
-            isBiometricsAuthButtonVisible={!isBiometricsAuthenticationAllowed}
-            onBiometricAuthPress={requestAuthenticationCheck}
-        />
+    const dispatch = useDispatch();
+
+    const biometricsError = useSelector(selectBiometricsError);
+    const shouldUserBeAuthenticated = useSelector(selectShouldUserBeAuthenticated);
+    const isBiometricsOptionEnabled = useSelector(selectIsBiometricsEnabled);
+
+    const handleBiometricsAppStateChange = useCallback(
```
Comment (peter-sanderson) [inline suite-native/biometrics/src/components/BiometricsModalRenderer.tsx]: nit: i would not bother with `useCallback` and simply put it into `useEffect` directly. 

This is the sole purpose of that `useEffect` hook ir map `addEventListener` subscription onto business logic. I see no point of spliting it

Comment (juriczech) [reply suite-native/biometrics/src/components/BiometricsModalRenderer.tsx]: It's used twice in the `useEffect`, so having it as a callback makes sense and as a rule-of-hooks & no-use-before-define, it needs to be `useCallback`

File: suite-native/storage/src/tests/migrations/biometricsV1.test.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    });
+    const createPersistedState = (overrides = {}) => ({
+        _persist: {
+            version: 1,
+            rehydrated: true,
+        },
+        ...overrides,
+    });
+
+    it('should migrate true value from storage to Redux state', () => {
```
Comment (peter-sanderson) [inline suite-native/storage/src/tests/migrations/biometricsV1.test.ts]: nit: I think we do ommit should 
```suggestion
    it('migrates true value from storage to Redux state', () => {
```

File: suite-native/biometrics/src/biometricsThunks.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    selectShouldUserBeAuthenticated,
+} from './biometricsSelectors';
+import { getIsBiometricsFeatureAvailable, getShouldRevokeAuth } from './biometricsUtils';
+import { AuthenticateError } from './types';
+
+export enum BiometricsToggleResult {
+    Enabled = 'enabled',
+    Disabled = 'disabled',
+    BiometricsNotAvailable = 'biometrics-not-available',
+}
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsThunks.ts:23]: Tiny nit, just personal preference of string unions over enum.
Feel free to close this:
```suggestion
export type BiometricsToggleResult = 'enabled' | 'disabled' | 'biometrics-not-available';
```

File: suite-native/biometrics/src/biometricsThunks.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+export const authenticateUserThunk = createThunk<
+    LocalAuthenticationResult,
+    void,
+    { rejectValue: AuthenticateError }
+>(`${BIOMETRICS_THUNK_PREFIX}/authenticate`, async (_, { rejectWithValue }) => {
+    const isBiometricsAvailable = await getIsBiometricsFeatureAvailable();
+
+    if (!isBiometricsAvailable) {
+        return rejectWithValue(AuthenticateError.BiometricsNotAvailable);
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsThunks.ts:51]: much better than previous code, where `authenticate` simply returned nothing 👍

File: suite-native/biometrics/src/biometricsSlice.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            .addCase(authenticateUserThunk.pending, state => {
+                state.biometricsError = null;
+            })
+            .addCase(authenticateUserThunk.fulfilled, state => {
+                state.isUserAuthenticated = true;
+                state.biometricsError = null;
+            })
+            .addCase(authenticateUserThunk.rejected, (state, { payload }) => {
+                if (payload === AuthenticateError.AuthenticationFailed) {
+                    state.biometricsError = payload ?? null;
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsSlice.ts]: can't be nullish inside the condition
```suggestion
                    state.biometricsError = payload;
```

File: suite-native/biometrics/src/biometricsSlice.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                state.isTogglingBiometricsSettingsOption = true;
+            })
+            .addCase(toggleBiometricsSettingsThunk.fulfilled, (state, { payload }) => {
+                if (payload === BiometricsToggleResult.Enabled) {
+                    state.isBiometricsEnabled = true;
+                }
+
+                if (payload === BiometricsToggleResult.Disabled) {
+                    state.isBiometricsEnabled = false;
+                }
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsSlice.ts]: I'd add comment `// BiometricsToggleResult.BiometricsNotAvailable is not considered a state change, but an unsuccessful execution, and only emits a warning alert.`
to make it clear why it is not exhaustive

File: suite-native/biometrics/src/components/BiometricOverlay.tsx [hunk 8]
```diff
@@ -25,28 +28,34 @@ const bottomWrapperStyle = prepareNativeStyle(utils => ({
...
-}: BiometricOverlayProps) => {
+export const BiometricOverlay = ({ isBiometricsAuthButtonVisible }: BiometricOverlayProps) => {
     const { applyStyle } = useNativeStyles();
 
+    const dispatch = useDispatch();
+
+    const shouldUserBeAuthenticated = useSelector(selectShouldUserBeAuthenticated);
+
+    const triggerManualAuthentication = () => {
+        if (shouldUserBeAuthenticated) {
```
Comment (Lemonexe) [inline suite-native/biometrics/src/components/BiometricOverlay.tsx]: `shouldUserBeAuthenticated` is always true, because `BiometricOverlay` is only rendered when true.
And if it were false, we have broken UI (button doesn't do anything).
In that case I'd omit the selector completely here, it only adds confusion.
Let's make it an assumption here in `BiometricOverlay` that it is only rendered when true.

File: suite-native/biometrics/src/biometricsThunks.ts [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        return rejectWithValue(AuthenticateError.AuthenticationFailed);
+    }
+
+    return result;
+});
+
+export const toggleBiometricsSettingsThunk = createThunk<
+    BiometricsToggleResult,
+    void,
+    { rejectValue: BiometricsToggleResult | AuthenticateError }
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsThunks.ts]: Nit: another case for string union.
In reject value you only need `BiometricsNotAvailable`, but enum does not easily let you extract a portion of it. So now it is typed that rejection can be also of type `Enabled` or `Disabled`, which can't happen.

If `BiometricsToggleResult` and `AuthenticateError` were string unions, you could type
`rejectValue: 'biometrics-not-available' | AuthenticateError`.

And vice-versa: fullfill cannot be of type `BiometricsToggleResult` but type interface makes it seems so.
That's what I was confused about in [my comment in slice](https://github.com/trezor/trezor-suite/pull/23224/changes#r3227788177), now I see it is meant that way, but if it was precisely typed, the slice matcher would indeed be exhaustive = clear.

File: suite-native/biometrics/src/components/BiometricsModalRenderer.tsx (hunk 2 above)
Comment (Lemonexe) [reply suite-native/biometrics/src/components/BiometricsModalRenderer.tsx]: @juriczech true. But I think Peter meant: define the function locally **within** the useEffect?
Current code is correct, but potentially dangerous: if you would change `handleBiometricsAppStateChange` – introduce a new dependency, it would cause useEffect to recalculate, not noticeable at first glance.
Inside useEffect this would of course be the same, but more easily noticeable that "oh, I need to add a new dependency to useEffect? Then I need to do things differently."

File: suite-native/biometrics/src/components/BiometricsModalRenderer.tsx [hunk 10]
```diff
@@ -1,15 +1,48 @@
...
+        }
+
+        handleBiometricsAppStateChange(AppState.currentState);
+
+        const subscription = AppState.addEventListener('change', handleBiometricsAppStateChange);
+
+        return () => {
+            subscription.remove();
+        };
+    }, [dispatch, handleBiometricsAppStateChange, isBiometricsOptionEnabled]);
```
Comment (Lemonexe) [inline suite-native/biometrics/src/components/BiometricsModalRenderer.tsx]: `dispatch` is redundant dependency.. But if you move the callback inside here, it won't be 😄 

Also, TIL that eslint `react-hooks/exhaustive-deps` does not mind unused dependencies??

File: suite-native/biometrics/src/biometricsThunks.ts [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+            payload: { enabled: true, origin: 'settingsToggle' },
+        });
+
+        return BiometricsToggleResult.Enabled;
+    },
+);
+
+export const handleBiometricsAppStateChangeThunk = createThunk<
+    BiometricsAppStateChangePayload | undefined,
+    { currentAppState: AppStateStatus },
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsThunks.ts]: this is rather `nextAppState`, isn't it?

File: suite-native/biometrics/src/biometricsThunks.ts [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+    { rejectValue: BiometricsAppStateChangeRejectReason }
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsThunks.ts:106]: note that the `rejectValue` is not handled, it's just dropped. It wasn't handled before either.

File: suite-native/biometrics/src/biometricsThunks.ts [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+        const isBiometricsOptionEnabled = selectIsBiometricsEnabled(getState());
+        const shouldRevokeAuth = getShouldRevokeAuth(goneToBackgroundAtTimestamp);
+
+        if (!isBiometricsOptionEnabled) {
+            return rejectWithValue('biometrics-disabled');
+        }
+
+        switch (currentAppState) {
+            case 'active':
+                if (goneToBackgroundAtTimestamp !== null && !shouldRevokeAuth) {
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsThunks.ts]: Can't you change `shouldRevokeAuth` to return `true` when `goneToBackgroundAtTimestamp` is `null`? 
It's not used anywhere else.
Then you can simplify this condition:
```suggestion
                if (!shouldRevokeAuth) {
```

File: suite-native/biometrics/src/biometricsThunks.ts [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+                }
+
+                return {
+                    isUserAuthenticated: false,
+                    goneToBackgroundAtTimestamp: Date.now(),
+                };
+
+            default:
+                return rejectWithValue('unhandled-app-state');
+        }
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsThunks.ts:159]: about the whole switch: I'm quite confused, because this is not 1:1 implementation.
There are some changes:
1. when changing to active and not within timelimit, previously it only invalidated authentication if you went from `background` to `active`, now also from `inactive` to `active`. I don't really know what it means.
2. `goneToBackgroundAtTimestamp` was only set when going to background, now also going `inactive`

Both changes SGTM, it makes more sense then before, but I don't understand details and  wonder if they are intended.

File: suite-native/biometrics/src/biometricsThunks.ts [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+                return {
+                    isUserAuthenticated: false,
+                    goneToBackgroundAtTimestamp: Date.now(),
+                };
+
+            case 'inactive':
+                if (isTogglingBiometricsInProgress) {
+                    return { goneToBackgroundAtTimestamp: Date.now() };
+                }
```
Comment (Lemonexe) [inline suite-native/biometrics/src/biometricsThunks.ts:150]: I don't understand why this condition at all.
If it goes inactive during toggling biometrics, I think this thunk shouldn't care, and set `isUserAuthenticated: false`. If toggling finishes, then biometrics is turned off and it doesn't matter.
If toggling fails, then we are left authenticated when we shouldn't.
Maybe what I'm describing is an ultra edge case, but I'm trying to understand why is this code written 🙂

File: suite-native/biometrics/src/__tests__/biometricsThunks.test.ts [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+                expect(store.getState().biometrics.isBiometricsEnabled).toBe(
+                    expectedIsBiometricsEnabled,
+                );
+                expect(analyticsReport).toHaveBeenCalledWith({
+                    type: events.biometricsChangeEvent.name,
+                    payload: {
+                        enabled: expectedIsBiometricsEnabled,
+                        origin: 'settingsToggle',
+                    },
+                });
```
Comment (Lemonexe) [inline suite-native/biometrics/src/__tests__/biometricsThunks.test.ts:150]: I would add
`expect(LocalAuthentication.authenticateAsync).toHaveBeenCalled();`
I think it's an important specification that toggling is only possible if it prompts you for the biometrics check.

**Review by Lemonexe — COMMENTED** [review]
Mostly LGTM, but I'd need a sync to understand some of the changes.

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 15 above)
Comment (Lemonexe) [reply suite-native/biometrics/src/biometricsThunks.ts:150]: explained during call, pls add comemnt :pray:

File: suite-native/biometrics/src/components/BiometricsModalRenderer.tsx (hunk 10 above)
Comment (juriczech) [reply suite-native/biometrics/src/components/BiometricsModalRenderer.tsx]: That's by design I think, but very unfortunate design IMO

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 11 above)
Comment (juriczech) [reply suite-native/biometrics/src/biometricsThunks.ts]: Speculative, it's `currentAppState` and `previousAppState` IMO 😄 , but let's stick to RN naming, which uses `nextAppState`. I'll rename it.

File: suite-native/storage/src/migrations/biometrics/v1.ts (hunk 1 above)
Comment (Lemonexe) [reply suite-native/storage/src/migrations/biometrics/v1.ts:36]: Still relevant.

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 11 above)
Comment (Lemonexe) [reply suite-native/biometrics/src/biometricsThunks.ts]: it is the newly changed appState so I don't think it's previous.. But I'd like `nextAppState` mainly for the consistency with other interfaces :slightly_smiling_face:

File: suite-native/biometrics/src/components/BiometricsModalRenderer.tsx (hunk 2 above)
Comment (juriczech) [reply suite-native/biometrics/src/components/BiometricsModalRenderer.tsx]: But also, then the useEffect becomes more expensive to rerun. 

However these are all microoptimizations so since you both seem to prefer to place it in the useEffect, I will do that...

Comment (juriczech) [reply suite-native/biometrics/src/components/BiometricsModalRenderer.tsx]: https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/storage/src/tests/migrations/biometricsV1.test.ts (hunk 3 above)
Comment (juriczech) [reply suite-native/storage/src/tests/migrations/biometricsV1.test.ts]: https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/biometricsSlice.ts (hunk 6 above)
Comment (juriczech) [reply suite-native/biometrics/src/biometricsSlice.ts]: Nice simplification, thanks!
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/biometricsSlice.ts (hunk 7 above)
Comment (juriczech) [reply suite-native/biometrics/src/biometricsSlice.ts]: After simplified return type, this was simplified even further, which would make the comment redundant IMO: https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/components/BiometricOverlay.tsx (hunk 8 above)
Comment (juriczech) [reply suite-native/biometrics/src/components/BiometricOverlay.tsx]: https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 9 above)
Comment (juriczech) [reply suite-native/biometrics/src/biometricsThunks.ts]: We can use `Extract<>` on enums too. If this is good enough for you, I'd keep it. If not, let's go for string union.
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/components/BiometricsModalRenderer.tsx (hunk 10 above)
Comment (juriczech) [reply suite-native/biometrics/src/components/BiometricsModalRenderer.tsx]: https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 11 above)
Comment (juriczech) [reply suite-native/biometrics/src/biometricsThunks.ts]: https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 12 above)
Comment (juriczech) [reply suite-native/biometrics/src/biometricsThunks.ts:106]: Yes, I kept previous behavior. Would be nice to handle this better, but probably keep it as a follow up, if even that.

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 13 above)
Comment (juriczech) [reply suite-native/biometrics/src/biometricsThunks.ts]: Great point. That simplifies things nicely!
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 14 above)
Comment (juriczech) [reply suite-native/biometrics/src/biometricsThunks.ts:159]: Yeah, all intended. Previous version was super unintuitive and a hell to read through IMHO. I tried to simplify it and with your comments I believe it's even better now. 

The absent 1:1 in this case is one of the main reasons for this refactor, but of course, worth mentioning and discussing.

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 15 above)
Comment (juriczech) [reply suite-native/biometrics/src/biometricsThunks.ts:150]: Comment added.
https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/__tests__/biometricsThunks.test.ts (hunk 16 above)
Comment (juriczech) [reply suite-native/biometrics/src/__tests__/biometricsThunks.test.ts:150]: https://github.com/trezor/trezor-suite/pull/23224/changes/6d9f4e0c0a12ff1937078bc569071d1b46d4c83d

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 12 above)
Comment (Lemonexe) [reply suite-native/biometrics/src/biometricsThunks.ts:106]: oki then :+1:

File: suite-native/biometrics/src/biometricsThunks.ts (hunk 9 above)
Comment (Lemonexe) [reply suite-native/biometrics/src/biometricsThunks.ts]: Perfect. Enums are much smarter than I thought :clap:

**Review by Lemonexe — APPROVED** [review]
Looks great now!

**Comment by Lemonexe** [comment]
All failing E2E related to Cardano, and failing in other runs too, so I'd ignore this – likely backend problems, it will probably get autoquarantined soon.

**Comment by honzauher** [comment]
**QA OK** now

- Lock/unlock → 30s rule, behaves correctly
- Cold start → biometrics prompt fires as expected, no silent auth
- Settings toggle → persists across restart
- Onboarding with biometrics → set correctly
- **App stays locked after going to background / foreground**

Info:
- Android 15
- develop-26.6.1, commit 6d88363c311b236dd6f7916ea47f281fa5ecfdfe
